### ISSUE TITLE
Feature: Make client configuration optional

### ISSUE THREAD

**Issue by JiriOndrusek** [quarkiverse/quarkiverse-minio#22]
It would be very helpful,  if configuration to create minio client is optional.

Let me describe use case:
I was implementing camel-quarkus minio extension by using quarkiverse-minio (apache/camel-quarkus#2116).
Camel-quarkus minio extension is capable of using several minio client at once or of defining clients dynamically in runtime.
It would be much easier and consistent if there was no mandatory client creation based on configuration.

In other words, quarkiverse-minio could be used to introduce minio capabilities, but client creation can be left to custom implementation.

**Comment by jtama-op** [comment]
I don't think I will completely remove the client creation from the extension. 
But I guess we could do the following : 

- If no configuration is provided : Do nothing
- If one ore more servers are configured : Produce clients

You could still create clients yourself anyway.

**Comment by JiriOndrusek** [comment]
Your solution seems to be exactly what I need.

**Comment by jtama-op** [comment]
After a small search I will just make configuration optional as providing multiple client forces me to have runtime AND compile time config, which would neef a bit more effort.

So for the moment I will just make configuration optional.

If it's present, a MinioClient will be provided, if not nothing happens.

**Comment by jtama-op** [comment]
Released in 0.2.0

**Comment by JiriOndrusek** [comment]
@jtama-op I've tested fix in 0.2.0 and it seems, that there is still a problem with validation - validation is executed even if configuration is empty (https://github.com/quarkiverse/quarkiverse-minio/blob/master/runtime/src/main/java/io/quarkiverse/minio/client/MinioProducer.java#L19)
Which causes exception:
````
2021-01-13 09:38:57,805 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (executor-thread-1) HTTP Request to /minio/initBucket failed, error id: 27f9fd56-3c0e-4589-a343-f9277f72b2b4-1: org.jboss.resteasy.spi.UnhandledException: io.quarkus.runtime.configuration.ConfigurationException: "quarkus.minio.url" is mandatory and must be a valid url
	at org.jboss.resteasy.core.ExceptionHandler.handleApplicationException(ExceptionHandler.java:106)
	at org.jboss.resteasy.core.ExceptionHandler.handleException(ExceptionHandler.java:372)
	at org.jboss.resteasy.core.SynchronousDispatcher.writeException(SynchronousDispatcher.java:218)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:519)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:138)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.access$000(VertxRequestHandler.java:41)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler$1.run(VertxRequestHandler.java:93)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2415)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1452)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at java.base/java.lang.Thread.run(Thread.java:834)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: io.quarkus.runtime.configuration.ConfigurationException: "quarkus.minio.url" is mandatory and must be a valid url
	at io.quarkiverse.minio.client.MinioProducer.verifyUrl(MinioProducer.java:35)
	at io.quarkiverse.minio.client.MinioProducer.produceMinioClient(MinioProducer.java:25)
	at io.quarkiverse.minio.client.MinioProducer_ProducerMethod_produceMinioClient_a74f1d65447af9f0051b3615caa3aca9cac0ca25_Bean.create(MinioProducer_ProducerMethod_produceMinioClient_a74f1d65447af9f0051b3615caa3aca9cac0ca25_Bean.zig:163)
	at io.quarkiverse.minio.client.MinioProducer_ProducerMethod_produceMinioClient_a74f1d65447af9f0051b3615caa3aca9cac0ca25_Bean.create(MinioProducer_ProducerMethod_produceMinioClient_a74f1d65447af9f0051b3615caa3aca9cac0ca25_Bean.zig:194)
	at io.quarkus.arc.impl.AbstractSharedContext.createInstanceHandle(AbstractSharedContext.java:96)
	at io.quarkus.arc.impl.AbstractSharedContext.access$000(AbstractSharedContext.java:14)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:29)
	at io.quarkus.arc.impl.AbstractSharedContext$1.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.LazyValue.get(LazyValue.java:26)
	at io.quarkus.arc.impl.ComputingCache.computeIfAbsent(ComputingCache.java:69)
	at io.quarkus.arc.impl.AbstractSharedContext.get(AbstractSharedContext.java:26)
	at io.quarkus.arc.impl.ClientProxies.getApplicationScopedDelegate(ClientProxies.java:17)
	at io.quarkiverse.minio.client.MinioProducer_ProducerMethod_produceMinioClient_a74f1d65447af9f0051b3615caa3aca9cac0ca25_ClientProxy.arc$delegate(MinioProducer_ProducerMethod_produceMinioClient_a74f1d65447af9f0051b3615caa3aca9cac0ca25_ClientProxy.zig:85)
	at io.quarkiverse.minio.client.MinioProducer_ProducerMethod_produceMinioClient_a74f1d65447af9f0051b3615caa3aca9cac0ca25_ClientProxy.bucketExists(MinioProducer_ProducerMethod_produceMinioClient_a74f1d65447af9f0051b3615caa3aca9cac0ca25_ClientProxy.zig:3891)
	at org.apache.camel.quarkus.component.minio.it.MinioResource.initBucket(MinioResource.java:175)
	at org.apache.camel.quarkus.component.minio.it.MinioResource_ClientProxy.initBucket(MinioResource_ClientProxy.zig:177)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:643)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:507)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:457)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:459)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:419)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:393)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:68)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)

````

**Comment by jtama-op** [comment]
If you don't ask for a MinioClient bean, then producer should not be called.
I tested it without any configuration, initiating the configuration myself and it worked. 
Do you have a reproducer ?

**Comment by JiriOndrusek** [comment]
If I understand correctly, if there is an autowired field for minioClient and no configuration, it would fail?
Reason of my problem is probably here https://github.com/apache/camel/blob/master/components/camel-minio/src/main/java/org/apache/camel/component/minio/MinioConfiguration.java#L65, camel component tries to autowire client and if it is not there, it creates new one.

**Comment by JiriOndrusek** [comment]
There is a possible reason in my test - https://github.com/apache/camel-quarkus/pull/2116/files#diff-f375f93ee01366fb03917896cef88ebb23934e84b58068f59ac64d3d42fd6950R67

I'll verify what is going there... This inject could be certainly refactored.
I'll let you know if this is the reason of my problem

**Comment by jtama-op** [comment]
Yes. If you want a MinioClient to be injected, then, you MUST provide a valide configuration.

**Comment by JiriOndrusek** [comment]
@jtama-op I've validated fix with camel-quarkus and everythink seems to be correct. Issue could be closed. Thanks for the fix.
